### PR TITLE
refactor code and add unit test for all agent models

### PR DESCRIPTION
**Pull request by shaunahu**
Things have done in this PR:
- [ML/AI models into a consolidated system](https://github.com/aodn/backlog/issues/6589)
- [Unit test all major functions](https://github.com/aodn/backlog/issues/6097)

Things not involved in this ticket:
- [Store request and response in S3](https://github.com/aodn/backlog/issues/6663)
- Update ~readme~ and notebook
- [Develop a training agent for machine learning models (keyword classification and data delivery mode classification)](https://github.com/aodn/backlog/issues/6697), which involves separating the training process from the original pipeline code to allow for modular use and easier maintenance.
- Store OpenAI API key in a secure environment
- [Use a free LLM for local development](https://github.com/aodn/backlog/issues/6667)

In this way, there are only two routers for the APIs: `process_record` and `health`:
image
image

### CODE REVIEW

File: data_discovery_ai/utils/agent_tools.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    """
+    # https://huggingface.co/docs/transformers/v4.47.1/en/model_doc/bert#transformers.TFBertModel
+    tokenizer = AutoTokenizer.from_pretrained("bert-base-uncased")
+
+    # use in Tensorflow https://huggingface.co/google-bert/bert-base-uncased
+    model = TFBertModel.from_pretrained("bert-base-uncased")
+
+    # https://huggingface.co/docs/transformers/main_classes/tokenizer#transformers.PreTrainedTokenizer.__call__.return_tensors, set as 'tf' to return tensorflow tensor
+    inputs = tokenizer(text, return_tensors="tf", max_length=512, truncation=True)
+    outputs = model(inputs)
```
[the remark this answers is not in the document]
Comment (shaunahu) [reply data_discovery_ai/utils/agent_tools.py:51]: I will ignore this because in each API call, this function is only called once when processing textual request (for keyword classification and data delivery mode classification) into embeddings

File: data_discovery_ai/core/agents/baseAgent.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+    def make_decision(self, request: Dict[str, str]) -> bool:
+        """
+        Make decision based on the request. This should be overridden by subclasses.
+        """
+        return True
+
+    def response(self) -> Dict[str, str]:
+        """
+        Agent makes response. This should be overridden by subclasses.
```
[the remark this answers is not in the document]
Comment (shaunahu) [reply data_discovery_ai/core/agents/baseAgent.py:37]: changed to `execute(self)`

**Comment by utas-raymondng** [comment]
Can you help to copy sections "Commit" and "Branching name" of the README of https://github.com/aodn/aodn-portal-v2 to this project's readme? I see the branching name is diff

**Comment by shaunahu** [comment]
> Can you help to copy sections "Commit" and "Branching name" of the README of https://github.com/aodn/aodn-portal-v2 to this project's readme? I see the branching name is diff

These two sections have been added into the data-discovery-ai readme through this PR. I will follow these rules in the future PR and commit. Please let me know if I need to rename this branch or not (if it renamed, the current PR will be closed, I will raise a new one)

**Comment by utas-raymondng** [comment]
When I do this

```
{
    "selected_model":["description_formatting"],
    "title": "RV Investigator Voyage IN2015_V02 ADCP Data",
    "abstract": "This record describes the Acoustic Doppler Current Profiler (ADCP) data collected from the Marine National Facility RV Investigator voyage IN2015_V02, titled: \"Sustained monitoring of the East Australian Current: Mass, heat and freshwater transports.\" The voyage took place between May 15 and May 26, 2015 departing Sydney (NSW) and returning to Brisbane (QLD).\n\nData was collected during in2015_v02 for the duration of the voyage. Data was collected using VMDAS and post-processed using CODAS.\nBoth the OS75 and the OS150 ADCPs were operated for the duration of the voyage.\nDue to issues with VMDAS acquiring serial data, all navigation data was regenerated from Techsas acquired NMEA strings.\nSignificant work was conducted by the science party to investigate the correct ADCP settings, which involved various permutations of broadband vs narrowband modes, and different keel depths. The findings are summarised in the report ‘ADCP_Report_INV2015_V02_Wijffels_final.pdf, supplied as part of the processed dataset. Additional documentation is in the ‘docs_from_voyage’ folder. The summary from that report is as follows:\n• both ADCP Ocean Surveyors had transducer alignment errors. The current estimate of the angles to put into VMDAS are: OS150 = 44.35 ( EA= 4435 in config file); OS75 = 52.44 (EA=5244 in config file)\n• Both OS’s were losing comms every 20 mins to 3 hours, then autostarting Sometimes the loss of comms would permanently stop acquisition (not auto restart). Changing out the Serial-USB comms cable might have fixed this.\n• the OS150 has several issues in BroadBand Mode which degrades the data:\n1. It is badly affected by bubbles flowing under the ship. This is lessened with reduced shipspeed and lowering the drop keep. It never really disappears unless the ship is stopped. The first 1-5 bins need to be carefully screened for unrealistic shears. This affect is reduced in Narrow Band mode.\n2. It suffers from strong interference from the OS75 and the EK60.\n• The OS75 suffers much less from bubble problems. It does however suffer from interference from both the OS150 and more weakly the EK60 (needs checking).\n• It is clear from the echograms that the sADCP’s interfere with the EK60. This would have to either be edited out or prevented for quantitative use of the EK60 output.\nEach permutation of settings used was processed separately, and the files concatenated together post processing. Please see ‘ADCP_Report_INV2015_V02_Wijffels_final.pdf’ for details of different settings used.\n\nThis dataset was processed and then archived within the CSIRO Oceans and Atmosphere (O&A) Information and Data Centre (IDC) in Hobart."
}
````

I see this in the log

Error in calling OpenAI API: Error code: 429 - {'error': {'message': 'Your account is not active, please check your billing details on our website.', 'type': 'billing_not_active', 'param': None, 'code': 'billing_not_active'}}


But the RESTAPI still reply OK with the un-formatted description, I think server should reply with error so people knows about it instead of suppressing the error

**Comment by utas-raymondng** [comment]
I think there is a need to add on free LLM model for dev env and we can config it by dev in dev.

File: data_discovery_ai/core/agents/keywordClassificationAgent.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
+# The agent model for executing the keyword classification task
+import numpy as np
+import pandas as pd
+from typing import Dict, List, Any, Optional
+from tensorflow.keras.models import load_model
```
Comment (utas-raymondng) [inline data_discovery_ai/core/agents/keywordClassificationAgent.py:5]: image

Cannot find reference 'keras' in '__init__.py | __init__.py' 

Install keras nightly??

File: data_discovery_ai/core/agents/keywordClassificationAgent.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+        self.type = "keyword_classification"
+        self.config = ConfigUtil()
+        self.model_config = self.config.get_keyword_classification_config()
+
+        self.pretrained_model_name = self.model_config["pretrained_model"]
+
+    def set_required_fields(self, required_fields) -> None:
+        return super().set_required_fields(required_fields)
+
+    def is_valid_request(self, request) -> bool:
```
Comment (utas-raymondng) [inline data_discovery_ai/core/agents/keywordClassificationAgent.py:26]: Signature not match the basedAgent

File: data_discovery_ai/core/agents/keywordClassificationAgent.py [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+            predicted_keywords = self.get_predicted_keywords(
+                predicted_labels, selective_labels
+            )
+
+            logger.info(f"Keyword is being predicted by {self.type} agent")
+            return predicted_keywords
+        else:
+            return []
+
+    def replace_with_column_names(
```
Comment (utas-raymondng) [inline data_discovery_ai/core/agents/keywordClassificationAgent.py:89]: image

Expected type 'list[str]', got 'list[int | list[int]]' instead

File: data_discovery_ai/core/agents/keywordClassificationAgent.py [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+    def get_predicted_keywords(self, prediction: np.ndarray, labels: Dict):
+        """
+        Convert binary predictions to textual keywords.
+        Input:
+            prediction: np.ndarray. The predicted binary matrix.
+            labels: Dict. The predefiend keywords.
+        Output:
+            predicted_keywords: pd.Series. The predicted keywords for the given targets.
+        """
+        target_predicted = pd.DataFrame(prediction, columns=labels)
```
Comment (utas-raymondng) [inline data_discovery_ai/core/agents/keywordClassificationAgent.py:111]: Parameter columns of pandas.core.frame.DataFrame.__init__
columns: ExtensionArray | ndarray | Index | Series | SequenceNotStr | range | None
 
Column labels to use for resulting frame when data does not have them,
defaulting to RangeIndex(0, 1, 2, ..., n). If data contains column labels,
will perform column selection instead.

File: data_discovery_ai/core/agents/descriptionFormattingAgent.py [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+
+        # load api key from .env file
+        load_dotenv()
+        self.openai_api_key = os.getenv("OPENAI_API_KEY")
+        self.model_config = ConfigUtil().get_description_formatting_config()
+
+    def set_required_fields(self, required_fields: list) -> None:
+        super().set_required_fields(required_fields)
+
+    def is_valid_request(self, request: Dict[str, str]) -> bool:
```
Comment (utas-raymondng) [inline data_discovery_ai/core/agents/descriptionFormattingAgent.py:28]: Signature not match with baseAgent

**Review by utas-raymondng — CHANGES_REQUESTED** [review]
Too many those syntax issue like signature not match with baseAgent and return type not match. 

You can consider using webstorm where it can catch those error easily

File: data_discovery_ai/core/agents/keywordClassificationAgent.py (hunk 3 above)
Comment (shaunahu) [reply data_discovery_ai/core/agents/keywordClassificationAgent.py:5]: It is a [tensorflow IDE issue](https://github.com/tensorflow/tensorflow/issues/56231) reported by tensorflow version >2.0. It's not an error nor using keras-nightly. I have changed to use the original keras as `from keras.api.models import load_model`

File: data_discovery_ai/core/agents/keywordClassificationAgent.py (hunk 6 above)
Comment (shaunahu) [reply data_discovery_ai/core/agents/keywordClassificationAgent.py:111]: I understand this concern, the labels are stored in the resource when training the model. I have updated the health check so that to hint when don't have the resource.
